### PR TITLE
Add ability to pass docker config and included registry credentials to build pods' docker clients, test builds and pushes, and add dockerApi.[extraArgs|extraFiles] to help testing

### LINKED ISSUE #26: ci: improve our tests to also trigger a build and push to a registry
We have automation setup that test installs the helm chart and ensures pods enter a running state successfully. We should do a bit more.

- [x] Setup a container registry, and grant permissions to k3s that is setup by jupyterhub/action-k3s-helm to pull from it ([something like this](https://github.com/jupyterhub/binderhub/blob/036877ffdf0abfde7e84f3972c7d0478cf4f7cb2/.github/workflows/test.yml#L93-L119))
- [ ] Make chartpress push images to the container registry as well ([something like this](https://github.com/jupyterhub/binderhub/blob/036877ffdf0abfde7e84f3972c7d0478cf4f7cb2/.github/workflows/test.yml#L195-L200))

I think we can be inspired by binderhub's strategies, but that we should aim for a lower complexity:
- [z2jh](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/main/.github/workflows/test-chart.yaml)
- [binderhub](https://github.com/jupyterhub/binderhub/blob/036877ffdf0abfde7e84f3972c7d0478cf4f7cb2/.github/workflows/test.yml)

### PR DESCRIPTION
**Pull request by consideRatio**
With binderhub-service being a Helm chart deploying to k8s, we absolutely require the ability to push the built artifact to a container registry somewhere. This enables the chart to create a k8s Secret with such credentials in a docker config format that can be mounted to the build pods via `c.KubernetesBuildExecutor.push_secret`. Users of the binderhub-service chart doesn't have to configure that manually, its instead set automatically.

## `buildPodsRegistryCredentials`

`buildPodsRegistryCredentials` should be provided with `server`, `username`, and `password` for the container registry. For a GCP based artifact-registry, this can be for example...

```yaml
config:
  BinderHub:
    use_registry: true
    image_prefix: europe-west1-docker.pkg.dev/binderhub-service-development/binderhub-service/
buildPodsRegistryCredentials:
  server: https://europe-west1-docker.pkg.dev
  # This GCP ServiceAccount is configured with:
  # - roles/artifactregistry.createOnPushWriter
  #
  username: _json_key
  password: |
    {
      "type": "service_account",
      "project_id": "binderhub-service-development",
      "private_key_id": "3ca6cb965bf9d8a23ee07220c6bfcdf70a2b4234",
      "private_key": "CENSORED",
      "client_email": "artifact-registry-push@binderhub-service-development.iam.gserviceaccount.com",
      "client_id": "101939388990083667377",
      "auth_uri": "https://accounts.google.com/o/oauth2/auth",
      "token_uri": "https://oauth2.googleapis.com/token",
      "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
      "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/artifact-registry-push%40binderhub-service-development.iam.gserviceaccount.com",
      "universe_domain": "googleapis.com"
    }
```

## `buildPodsDockerConfig`

This may not be relevant for us to use, but was added for parity with binderhub chart to some degree.

```
buildPodsDockerConfig:
  someDockerConfigRecognizableKey: and-value
```

## `dockerApi.[extraArgs|extraFiles]`

In order to setup tests with build/push, I needed to configure the docker daemon runnin on the host node to tolerate interacting with HTTP based docker registries like the one we have. Instead of doing like in jupyterhub/binderhub where a configmap is created and mounted etc in a way that is very hard to follow, I do something that is just quite hard to follow by introducing `extraFiles` taken from z2jh.

## Tests build/push

- Closes #26 by testing build and push

### CODE REVIEW

File: dev-config.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# FIXME: When running tests, we will need a local container registry to test
+#        pushing images to that can be reached from the build pods.
+#
```
Comment (GeorgianaElena) [inline dev-config.yaml:3]: @consideRatio, is this the only bit that's currently missing from this PR?

Comment (consideRatio) [reply dev-config.yaml:3]: Its missing, but it doesn't have to be in this PR I figure!

I'll dig into this PR once again now this afternoon. I just finished with a new hub deployment and will eat some food.

Comment (GeorgianaElena) [reply dev-config.yaml:3]: No hurry! Thank you! let me know if there's anything I can help with

Comment (consideRatio) [reply dev-config.yaml]: We now test and push against a local registry!

**Comment by consideRatio** [comment]
@GeorgianaElena I think this is good to go finally, its quite a messy PR because it introduced dockerApi.extraFiles for example, but that is copy-pasted from z2jh where it has been very successfully used reliably to inject files into pods.

**Comment by GeorgianaElena** [comment]
Thanks @consideRatio!
